File: haskell/bob/bob.hs
module Bob (responseFor) where

import Data.Char

responseFor :: String -> String
responseFor x
  | all isSpace x = "Fine. Be that way!"
  | any isUpper x && not (any isLower x) = "Woah, chill out!"
  | last x == '?' = "Sure."
  | otherwise = "Whatever."

File: clojure/bob/bob.clj
(ns bob)

(defn- has-upper? [s]
  (some #(Character/isUpperCase %) s))

(defn- has-lower? [s]
  (some #(Character/isLowerCase %) s))

(defn- is-shouting? [s]
  (and (has-upper? s) (not (has-lower? s))))

(defn- is-question? [s]
  (= (last s) \?))

(defn response-for [s]
  (cond
    (clojure.string/blank? s) "Fine. Be that way!"
    (is-shouting? s) "Woah, chill out!"
    (is-question? s) "Sure."
    :else "Whatever."))
